File: Terminplaner-Ruby/src/Termin.rb
require "Teilnahme"

class Termin 

	include Comparable

	attr_reader :bezeichnung, :autor, :start_zeit, :dauer_in_minuten, :teilnahmen

	def initialize(autor, start_zeit, dauer_in_minuten, bezeichnung)
		@autor = autor
		@start_zeit = start_zeit
		@dauer_in_minuten = dauer_in_minuten
		@bezeichnung = bezeichnung
		@teilnahmen = Set.new
		@teilnahmen << Teilnahme.new(Teilnahme::BESTAETIGT, autor)
	end

	def <=> (obj)
		start_zeit_compare = @start_zeit <=> obj.start_zeit

		if start_zeit_compare == 0 then to_s <=> obj.to_s else start_zeit_compare end
	end

	def to_s 
		"#{start_zeit}: #{bezeichnung} (#{dauer_in_minuten} Min.)"
	end

	def lade_teilnehmer_ein(einTeilnehmer) 
		@teilnahmen << Teilnahme.new_default(einTeilnehmer)
	end

	def teilnehmer
		@teilnahmen.inject([]) { |result,teilnahme|
		  	result << teilnahme.teilnehmer
	    	}.sort
	end

	def teilnahme_status(benutzer)
		teilnahme(benutzer).status
	end

	def bestaetige_termin(benutzer)
		teilnahme(benutzer).bestaetige
	end

	def lehne_termin_ab(benutzer)
		teilnahme(benutzer).lehne_ab
	end

	def fuege_zu_terminliste_hinzu(terminliste, benutzer, zeige_auch_abgelehnte)
		benutzer_ok = benutzer == autor || ist_eingeladen?(benutzer)
		if benutzer_ok and (zeige_auch_abgelehnte or ist_teilnahme_noch_moeglich?(benutzer))
			terminliste << self 
		end
	end

	def teilnahme(benutzer)
		teilnahmen.each { |teilnahme|
			if benutzer == teilnahme.teilnehmer
				return teilnahme
		    	end
		}
		raise BenutzerNichtEingeladenException.new
	end

	def ist_teilnahme_noch_moeglich?(benutzer)
		return ist_eingeladen?(benutzer) && !teilnahme(benutzer).ist_abgelehnt
	end

	def ist_eingeladen?(benutzer) 
		begin
			teilnahme(benutzer)
			return true
		rescue BenutzerNichtEingeladenException
			return false
		end
	end

end

File: Terminplaner-Java/src/Termin.java
import java.util.ArrayList;
import java.util.Collection;
import java.util.Date;
import java.util.SortedSet;
import java.util.TreeSet;

public class Termin implements Comparable {

	private final String _autor;

	private Date _startZeit;

	private int _dauerInMinuten;

	private String _bezeichnung;

	private SortedSet<Teilnahme> _teilnahmen = new TreeSet<Teilnahme>();

	public Termin(String autor, Date startZeit, int dauerInMinuten,
			String bezeichnung) {
		_autor = autor;
		_startZeit = startZeit;
		_dauerInMinuten = dauerInMinuten;
		_bezeichnung = bezeichnung;
		_teilnahmen.add(new Teilnahme(TeilnahmeStatus.BESTAETIGT, autor));
	}

	public String getBezeichnung() {
		return _bezeichnung;
	}

	public int getDauerInMinuten() {
		return _dauerInMinuten;
	}

	public Date getStartZeit() {
		return _startZeit;
	}

	public int compareTo(Object o) {
		if (!(o instanceof Termin))
			return -1;

		int startZeitCompare = getStartZeit().compareTo(
				((Termin) o).getStartZeit());

		if (startZeitCompare == 0)
			return toString().compareTo(((Termin) o).toString());

		return startZeitCompare;
	}

	public String toString() {
		return getStartZeit() + ": " + getBezeichnung() + " ("
				+ getDauerInMinuten() + " Min.)";
	}

	public void ladeTeilnehmerEin(String teilnehmer) {
		_teilnahmen.add(new Teilnahme(teilnehmer));
	}

	public Collection<String> getTeilnehmer() {
		Collection<String> result = new ArrayList<String>();
		for (Teilnahme teilnahme : _teilnahmen) {
			result.add(teilnahme.getTeilnehmer());
		}
		return result;
	}

	public String getAutor() {
		return _autor;
	}

	public TeilnahmeStatus getTeilnahmeStatus(String benutzer)
			throws BenutzerNichtEingeladenException {
		return getTeilnahme(benutzer).getStatus();
	}

	public void bestaetigeTermin(String benutzer)
			throws BenutzerNichtEingeladenException {
		getTeilnahme(benutzer).bestaetige();
	}

	public void lehneTerminAb(String benutzer)
			throws BenutzerNichtEingeladenException {
		getTeilnahme(benutzer).lehneAb();
	}

	void fuegeZuTerminlisteHinzu(Collection<Termin> terminListe,
			String benutzer, boolean zeigeAuchAbgelehnte) {
		boolean benutzerOK = benutzer.equals(getAutor())
				|| _teilnahmen.contains(new Teilnahme(benutzer));
		if (benutzerOK && (zeigeAuchAbgelehnte || istTeilnahmeNochMoeglich(benutzer))) {
			terminListe.add(this);
		}
	}

	private Teilnahme getTeilnahme(String benutzer)
			throws BenutzerNichtEingeladenException {
		for (Teilnahme teilnahme : _teilnahmen) {
			if (teilnahme.getTeilnehmer().equals(benutzer))
				return teilnahme;
		}
		throw new BenutzerNichtEingeladenException();
	}

	private boolean istTeilnahmeNochMoeglich(String benutzer) {
		try {
			Teilnahme teilnahme = getTeilnahme(benutzer);
			return !teilnahme.istAbgelehnt();
		} catch (BenutzerNichtEingeladenException e) {
			return false;
		}
	}

}
